platforms/core-runtime/build-operations-trace/src/main/java/org/gradle/internal/operations/trace/BuildOperationTrace.java

```diff
@@ -605,4 +642,116 @@ public void finished(BuildOperationDescriptor buildOperation, OperationFinishEve
             }
         }
     }
+
+    interface TraceWriter {
+
+        /**
+         * Write a serialized operation to the log file.
+         */
+        void write(SerializedOperation serializedOperation);
+
+        /**
+         * This method must be called after all write operations have been completed.
+         */
+        void complete(boolean writeTree);
+
+    }
+
+    /**
+     * A {@link TraceWriter} that offloads all writing operations to a separate thread.
+     */
+    private static class AsyncTraceWriter implements TraceWriter {
+
+        private final TraceWriter delegate;
+        private final AsyncExecutor executor;
+
+        public AsyncTraceWriter(TraceWriter delegate) {
+            this.delegate = delegate;
+            this.executor = new AsyncExecutor();
+        }
+
+        @Override
+        public void write(SerializedOperation operation) {
+            executor.execute(() -> delegate.write(operation));
+        }
+
+        @Override
+        public void complete(boolean outputTree) {
+            try {
+                executor.execute(() -> delegate.complete(outputTree));
+            } finally {
+                IoActions.closeQuietly(executor);
+            }
+        }
+
+    }
+
+    /**
+     * Executes submitted operations sequentially in a separate thread.
+     * <p>
+     * This executor takes special care to ensure that any exceptions thrown by
+     * submitted actions are rethrown on the calling thread, rather than being
+     * silently ignored.
+     * <p>
+     * The use case for this executor strongly overlaps with that of
+     * {@link org.gradle.kotlin.dsl.concurrent.AsyncIOScopeFactory}.
+     * We should consider merging these implementations.
+     */
+    private static class AsyncExecutor implements Closeable {
+
+        private final ExecutorService executor;
+        private final AtomicReference<@Nullable Throwable> failure = new AtomicReference<>();
+
+        public AsyncExecutor() {
+            this.executor = Executors.newSingleThreadExecutor();
```

Comment:
This is an unbounded queue, which means that if build operations are reported too fast we may get an out-of-memory error. We should consider bounding this queue (via `ArrayBlockingQueue`) to a reasonable amount (1000?) so that we slow down threads instead of crashing.

Reply:
I filed this as https://github.com/gradle/gradle/issues/35009